### PR TITLE
Fix GPU's `gradient!` 

### PR DESCRIPTION
**Pull request by paraynaud**

### CODE REVIEW

**Comment by paraynaud** [comment]
Works on GPU too.

**Comment by paraynaud** [comment]
I will test the last commit on a GPU tomorrow.

**Review by farhadrclass — COMMENTED** [review]
I am good with this but should we have a unit test for it (When GPU available the unit test run, Still not sure if GPU is supported on unit test ? )

File: src/KnetNLPModels.jl [hunk 1]
```diff
@@ -77,17 +77,16 @@ function KnetNLPModel(
...
+  (xtst, ytst) = data_test
+   
   training_minibatch_iterator = create_minibatch(xtrn, ytrn, size_minibatch)
   test_minibatch_iterator = create_minibatch(xtst, ytst, size_minibatch)
   current_training_minibatch = rand(training_minibatch_iterator)
   current_test_minibatch = rand(test_minibatch_iterator)
 
   nested_array = build_nested_array_from_vec(chain_ANN, x0)
-  layers_g = similar(params(chain_ANN)) # create a Vector of layer variables
+  layers_g = similar.(params(chain_ANN)) # create a Vector of layer variables
```
[the remark this answers is not in the document]
Comment (farhadrclass) [reply src/KnetNLPModels.jl:89]: One thing I am not sure is how to test on GPUin unit test as of now, so I wanted to double check if it is possible

[the remark this answers is not in the document]
Comment (paraynaud) [reply src/KnetNLPModels.jl:89]: > why is param here lower case and in the other [src/KnetNLPModels_methods.jl](https://github.com/JuliaSmoothOptimizers/KnetNLPModels.jl/pull/74/files#diff-670c5d3f5eca9e615b8e0927c7d2f89c16e44b4c8acd4c477f860166e3df3344) it is Param

`params` is a method and not the type.
The method return a type `Param` if needed.

> One thing I am not sure is how to test on GPUin unit test as of now, so I wanted to double check if it is possible

I pull on the GPU the branch I want to test, and I run manually the tests on the GPU.
Which is nice with KnetNLPModels.jl is that the code works for both CPU and GPU, so i just run `] test` once julia in loaded on Atlas.

[the remark this answers is not in the document]
Comment (farhadrclass) [reply src/KnetNLPModels.jl:89]: so the test run both GPU and CPU then

**Comment by farhadrclass** [comment]
> 

Can we open a bug here and put it conditional on knet.jl update?
I will approve this branch and we push it with a bug to fix in the future

**Comment by farhadrclass** [comment]
And I think you mentioned this buildkite/knetnlpmodels-dot-jl fails,
Can we remove it and add it to the bug or fix it so it doesn't fail?

**Comment by paraynaud** [comment]
> > 
> 
> Can we open a bug here and put it conditional on knet.jl update? I will approve this branch and we push it with a bug to fix in the future

I don't understand what you mean.
Does bug == github issue ?

I prefer keeping buildkite/knetnlpmodels-dot-jl even if it fails, because Dominique had to ask for us to get access to GPU from buildkite's community.
I'm not sure how it'll react if we remove buildkite/knetnlpmodels-dot-jl.

**Comment by farhadrclass** [comment]
Sorry for confusion 
yes bug is github issue, 

Sure if we already expect it to fail and have it in the github issue then I am okay with it 

I will pass approve it then, just don''t forget the issue

**Review by farhadrclass — APPROVED** [review]

**Comment by paraynaud** [comment]
#75

**Pull request merged**